src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs

```diff
@@ -120,30 +125,17 @@ internal MetadataDb(int payloadLength)
                     initialSize = OneMegabyte;
                 }
 
-                _data = ArrayPool<byte>.Shared.Rent(initialSize);
-                Length = 0;
-#if DEBUG
-                _isLocked = false;
-#endif
+                byte[] data = ArrayPool<byte>.Shared.Rent(initialSize);
+                return new MetadataDb(data, isLocked: false, convertToAlloc);
             }
 
-            internal MetadataDb(MetadataDb source, bool useArrayPools)
+            internal static MetadataDb CreateLocked(int payloadLength)
             {
-                Length = source.Length;
-
-#if DEBUG
-                _isLocked = !useArrayPools;
-#endif
+                // Add one row worth of data since we need at least one row for a primitive type.
+                int size = payloadLength + DbRow.Size;
 
-                if (useArrayPools)
-                {
-                    _data = ArrayPool<byte>.Shared.Rent(Length);
-                    source._data.AsSpan(0, Length).CopyTo(_data);
-                }
-                else
-                {
-                    _data = source._data.AsSpan(0, Length).ToArray();
-                }
+                byte[] data = new byte[size];
```

Comment:
How large can `size` be here, in the worst case, and how come we can get away with using GC-backed array? `JsonDocument` APIs are meant to be amortized non-allocating for payloads under 2 MB.

Reply:
Much of the optimizations in this PR are due to not using the pooled buffer, since it was previously created and then converted to a normal alloc in "Clone()" and in all cases, except object and array, using the pooled buffer was not necessary.

`size` can be the size of the largest JSON value (utf8Json.Length).

Reply:
> `size` can be the size of the largest JSON value (utf8Json.Length).

And does this **only** happen in similar code paths where previously `Clone()` would be called?

**To confirm:** Are we still non-allocating for the standard use of JsonDocument/JsonElement parse, random-access, and dispose, or does this code path come into play in that usage in some way?

Reply:
> Are we still non-allocating for the standard use of JsonDocument/JsonElement parse, random-access, and dispose, or does this code path come into play in that usage in some way?

Yes; the intent was to use the pooled memory exactly the same for the standard usage. Also I just verified that the "before" and "after" results have the same values for the "Allocated" column.